### ISSUE TITLE
Missing __init__.py causes inconsistent failure and confusing error message

### ISSUE THREAD

**Issue by intgr** [python/mypy#12104]
**Bug Report**

With a folder structure like this:

```
module
├── __init__.py
└── submodule
    ├── code.py
    └── types.py
```

**Actual Behavior**

Running mypy directly on `code.py` causes this error output:
```
% mypy module/submodule/code.py
mypy: "module/submodule/types.py" shadows library module "types"
note: A user-defined top-level module with name "types" is not supported
```

However, running mypy on the folders is successful and propery type-checks the code, no errors or warnings.
```
% mypy module/
Success: no issues found in 3 source files
% mypy module/submodule/
Success: no issues found in 2 source files
```

When using dmypy, the behavior is inconsistent and depends on previous invocations.
```
% dmypy run module
Success: no issues found in 3 source files
% dmypy run module/submodule/code.py
Success: no issues found in 1 source file
% dmypy stop
Daemon stopped
% dmypy run module/submodule/code.py
Daemon started
mypy: "module/submodule/types.py" shadows library module "types"
note: A user-defined top-level module with name "types" is not supported
```

**Expected Behavior**

1. If the missing `__init__.py` file or existence of `types.py` module is an error, it should be an error always, regardless of whether I invoke mypy with file path or directory path.
2. The error message would be more helpful if it indicated the missing `__init__.py`

**Your Environment**

- Mypy version used: 0.931
- Mypy command-line flags: -
- Mypy configuration options from `mypy.ini` (and other config files): -
- Python version used: 3.10.1
- Operating system and version: macOS 12.1

**Comment by hauntsaninja** [comment]
This works if you set `--namespace-packages`, which you should in general if you want things to work without `__init__.py`. We plan on making this the default soon, which will help solve several similar bad surprises.
